### ISSUE TITLE
Batch execute not running

### ISSUE THREAD

**Issue by simkessy** [pnp/pnpjs#574]
### Category
- [ ] Enhancement
- [ ] Bug
- [x] Question
- [ ] Documentation gap/issue

### Version

Please specify what version of the library you are using: ["@pnp/pnpjs": "^1.3.0"]

Please specify what version(s) of SharePoint you are targeting: [    Online     ]

I have a batch call that's not running, not throwing any errors and not creating the `batch` call in the network tab after I call execute on it. I'm not sure how to debug from here. 

This is what the object looks like
![image](https://user-images.githubusercontent.com/5871660/54792305-c3c6be80-4bfa-11e9-926f-d8916991438c.png)

**Comment by patrick-rodgers** [comment]
Can you please share your code?

**Comment by simkessy** [comment]
Hey @patrick-rodgers 

I put together this script to create lists from a collection. So there's quite a bit there. 
```
var deploy = deploy || {};

Object.assign(deploy, {
    // Global error handling
    error: e => {
        console.log(e);
    },

    // Set up list for reuse
    setup: async () => {
        let promises = deploy.config.lists.map(async list => {
            list.batch = pnp.sp.web.createBatch();
            list.instance = pnp.sp.web.lists.getByTitle(list.displayName);
            list.fields = list.instance.fields.inBatch(list.batch);

            return list.instance
                .getListItemEntityTypeFullName()
                .then(entity => {
                    list.entity = entity;
                });
        });

        return await Promise.all(promises);
    },

    // Manage list functions
    lists: {
        // ------------- Create lists -------------//
        create: async () => {
            let batch = pnp.sp.web.createBatch();

            console.log("create batch:", batch);

            deploy.config.lists.map(list => {
                pnp.sp.web.lists
                    .inBatch(batch)
                    .add(
                        list.internalName,
                        list.description,
                        list.template,
                        true,
                        list.additionalSettings
                    )
                    .then(res => {
                        // Store created list guid
                        list.guid = res.data.Id;
                        console.debug(`Created list: ${list.internalName}`);
                    })
                    .catch(error => {
                        console.debug("error");
                        deploy.error(error);
                    });
            });

            try {
                return await batch.execute();
            } catch (err) {
                console.log(err);
            }
        },

        // ------------- Update List display name -------------//
        update: async () => {
            let batch = pnp.sp.web.createBatch();

            console.log("update batch:", batch);

            deploy.config.lists.map(list => {
                pnp.sp.web.lists
                    .getByTitle(list.internalName)
                    .inBatch(batch)
                    .update({
                        Title: list.displayName
                    })
                    .then(() => {
                        console.debug(
                            `List: Renamed ${list.internalName} to ${
                                list.displayName
                            }`
                        );
                    })
                    .catch(error => {
                        console.debug("error");
                        deploy.error(error);
                    });
            });

            try {
                return await batch.execute();
            } catch (e) {
                console.log(e);
            }
        },

        // ------------- Delete Lists -------------//
        delete: async () => {
            let promises = deploy.config.lists.map(async list => {
                try {
                    return await pnp.sp.web.lists
                        .getByTitle(list.displayName)
                        .delete()
                        .then(() => {
                            console.debug(`Deleted List: ${list.displayName}`);
                        });
                } catch (error) {
                    console.debug(`List does not exist: ${list.displayName}`);

                    // console.debug(deploy.error(error));
                }
            });

            return await Promise.all(promises);
        },

        // ------------- Seed Lists -------------//
        seed: () => {
            deploy.config.lists.forEach((list, index) => {
                let buildItem = () => {
                    let list = deploy.config.lists[index];
                    let columns = list.columns;
                    let seeds = list.seeds;

                    let itemData = {};

                    columns.map((col, index) => {
                        let columnName =
                            col.fieldType === "Lookup"
                                ? `${col.internalName}Id`
                                : col.internalName;
                        let seed = seeds[index];

                        if (seed !== null) {
                            seed =
                                col.fieldType === "Text"
                                    ? String(seed())
                                    : seed();
                            itemData[columnName] = seed;
                        }
                    });

                    // console.debug(
                    //     "List",
                    //     list.internalName,
                    //     "itemData",
                    //     itemData
                    // );

                    return itemData;
                };

                // Create random items
                Array(faker.random.number({ min: 10, max: 15 }))
                    .fill()
                    .map((val, index) => {
                        list.instance.items
                            .inBatch(list.batch)
                            .add(buildItem(), list.entity)
                            .then(() => {
                                // console.debug(
                                //     `DUMMY DATA - ${
                                //         list.displayName
                                //     } - ${index}`
                                // );
                            })
                            .catch(err => {
                                console.debug("err", err);
                                console.debug(
                                    `ERROR DUMMY DATA - ${
                                        list.displayName
                                    } - ${index}`
                                );
                            });
                    });
            });
        }
    },

    // Manage field functions
    fields: {
        define: (internalName, displayName, fieldType, properties = []) => ({
            internalName,
            displayName,
            fieldType,
            properties
        }),

        // Find relevant function based on field type
        mapper: {
            // title: string, maxLength?: number, properties?: FieldCreationProperties
            Text: (fields, title, props = []) =>
                fields.addText(title, ...props),

            // title: string, numberOfLines?: number, richText?: boolean, restrictedMode?: boolean, appendOnly?: boolean, allowHyperlink?: boolean, properties?: FieldCreationProperties
            MultiText: (fields, title, props = []) =>
                fields.addMultilineText(title, ...props),

            // title: string, choices: string[], format?: ChoiceFieldFormatType, fillIn?: boolean, properties?: FieldCreationProperties
            Choice: (fields, title, props = []) =>
                fields.addChoice(title, ...props),

            // title: string, choices: string[], fillIn?: boolean, properties?: FieldCreationProperties
            MultiChoice: (fields, title, props = []) =>
                fields.addMultiChoice(title, ...props),

            // title: string, selectionMode: FieldUserSelectionMode, properties?: FieldCreationProperties
            User: (fields, title, props = []) =>
                fields.addUser(title, ...props),

            // title: string, displayFormat?: DateTimeFieldFormatType, calendarType?: CalendarType, friendlyDisplayFormat?: number, properties?: FieldCreationProperties
            Date: (fields, title, props = []) =>
                fields.addDateTime(title, ...props),

            // title: string, formula: string, dateFormat: DateTimeFieldFormatType, outputType?: FieldTypes, properties?: FieldCreationProperties
            Calculated: (fields, title, props = []) =>
                fields.addCalculated(title, ...props),

            // title: string, minValue?: number, maxValue?: number, properties?: FieldCreationProperties
            Number: (fields, title, props = []) =>
                fields.addNumber(title, ...props),

            // title: string, minValue?: number, maxValue?: number, currencyLocalId?: number, properties?: FieldCreationProperties
            Currency: (fields, title, props = []) =>
                fields.addCurrency(title, ...props),

            // title: string, displayFormat?: UrlFieldFormatType, properties?: FieldCreationProperties
            Url: (fields, title, props = []) => fields.addUrl(title, ...props),

            // title: string, lookupListId: string, lookupFieldName: string, properties?: FieldCreationProperties
            Lookup: (fields, title, props) => {
                let propsClone = _.clone(props, true);
                let list = propsClone.shift()();

                return fields.addLookup(title, list.guid, ...propsClone);
            },

            // title: string, properties?: FieldCreationProperties
            Boolean: (fields, title, props = []) =>
                fields.addBoolean(title, ...props),

            // XML
            XML: (fields, title, props = []) =>
                fields.createFieldAsXm(title, ...props)
        },

        // Create field
        add: () => {
            // Create fields for each list in batch
            deploy.config.lists.map(list => {
                list.columns.map(col => {
                    return deploy.fields.mapper[col.fieldType](
                        list.fields,
                        col.internalName,
                        col.properties
                    )
                        .then(() => {
                            console.debug(
                                `Field: Created ${col.internalName} in list: ${
                                    list.displayName
                                }`
                            );
                        })
                        .catch(error => {
                            console.debug("error");
                            deploy.error(error);
                        });
                });
            });
        },

        // Rename fields
        rename: () => {
            // Make calls in batch
            deploy.config.lists.map(list => {
                list.columns.map(col => {
                    // Test if col has display name
                    if (!col.hasOwnProperty("displayName")) return false;

                    list.fields
                        .getByInternalNameOrTitle(col.internalName)
                        .inBatch(list.batch)
                        .update({
                            Title: col.displayName
                        })
                        .then(() => {
                            console.debug(
                                `Field: Renamed ${col.internalName} to ${
                                    col.displayName
                                } in list: ${list.displayName}`
                            );
                        })
                        .catch(error => {
                            console.debug("error");
                            deploy.error(error);
                        });
                });
            });
        }
    },

    // Update default view with created fields
    views: {
        // ------------- Update View-------------//
        update: (target = "All Items") => {
            deploy.config.lists.map(list => {
                list.instance.views
                    .getByTitle(target)
                    .fields.inBatch(list.batch)
                    .removeAll()
                    .then(() => {
                        console.debug(`Reset view: ${list.displayName}`);
                    });

                list.columns.map(col => {
                    list.instance.views
                        .getByTitle(target)
                        .fields.inBatch(list.batch)
                        .add(col.internalName)
                        .then(() => {
                            console.debug(
                                `View: Updated ${target} - Added ${
                                    col.displayName
                                } in ${list.displayName}`
                            );
                        })
                        .catch(error => {
                            console.debug("error");
                            deploy.error(error);
                        });
                });
            });
        }
    },

    init: async (seed = false) => {
        console.clear();

        // let warning = confirm(
        //     "This will delete the lists if any exist and recreate them"
        // );

        if (true) {
            deploy.loading = SP.UI.ModalDialog.showWaitScreenWithNoClose(
                "Setting up app lists"
            );
            try {
                // Delete lists if they exist
                await deploy.lists.delete();

                // Create lists
                await deploy.lists.create();

                // Update list
                await deploy.lists.update();

                // Set up lists and pnp artifacts (batch, instance, fields)
                // await deploy.config.lists.map(deploy.setup);
                await deploy.setup();

                // Add fields to lists
                deploy.fields.add();

                // Rename fields
                // deploy.fields.rename();

                // Update allItems view
                // deploy.views.update();

                // Seed lists
                // if (seed) {
                //     // deploy.lists.seed();
                // }

                // // Remove required from title
                // deploy.config.lists.map(list => {
                //     pnp.sp.web.lists
                //         .getByTitle(list.displayName)
                //         .fields.getByTitle("Title")
                //         .inBatch(list.batch)
                //         .update({
                //             Required: false
                //         })
                //         .then(
                //             console.debug(
                //                 `Set title required false in list: ${
                //                     list.displayName
                //                 }`
                //             )
                //         );
                // });

                // Execute batches
                for (list of deploy.config.lists) {
                    console.log("list:", list.batch);
                    await list.batch
                        .execute()
                        .then(console.log)
                        .catch(e => {
                            console.log("wtf?", e);
                        });
                }

                console.debug("--------------------------------------");
                console.debug("SETUP COMPLETE, THANK YOU");
                console.debug("--------------------------------------");

                deploy.loading.close();

                // Display final results
                SP.UI.ModalDialog.showErrorDialog(
                    "Setup is complete",
                    `<br>
                    <h3>Created ${
                        deploy.config.lists.length
                    } lists:</h3><br> ${deploy.config.lists
                        .map(list => {
                            return `${list.displayName} <br>`;
                        })
                        .join("")}`
                );
            } catch (err) {
                deploy.loading.close();

                console.debug("--------------------------------------");
                console.debug("SETUP FAILED: Deleting lists...");
                console.log(err);
                console.debug("--------------------------------------");

                await deploy.lists.delete();

                // Set up lists and pnp artifacts (batch, instance, fields)
                deploy.config.lists.map(deploy.setup);

                console.debug("--------------------------------------");
                console.debug("LISTS DELETED, THANK YOU!");
                console.debug("--------------------------------------");
            }
        }
    }
});
```

List definition 
```
var deploy = deploy || {};

deploy.config = {};

Object.assign(deploy.config, {
    lists: [
        {
            internalName: "myList",
            displayName: "myList",
            description:
                "",
            template: 100,
            additionalSettings: {},
            columns: [
                deploy.fields.define(
                    "myColumn",
                    "myColumn",
                    "Text"
                ),
                deploy.fields.define("ProjectNumber", "Project Number", "Text")
            ]
        }

    ]
});
```

**Comment by simkessy** [comment]
Is there a difference between `execute` and `executeImpl'

![image](https://user-images.githubusercontent.com/5871660/54798122-52940500-4c14-11e9-96f3-93bacf374788.png)

executeImpl actually makes the call.

**Comment by patrick-rodgers** [comment]
I'll be honest, that is too much code to go through. Can you isolate the part that has errors or issues? Also, this may be a duplicate of the issue found in #564 so may be fixed by the same fix. executeImpl is the internal protected method that is the individual implementations of batch, for SP and Graph for example. Calling that directly would "work" in that a request would go out, but you would have skipped some of the steps invoked within execute and it is not recommended as it likely won't work.

**Comment by simkessy** [comment]
Okay, seems like the PR is still open however, any idea when it will be merged into main or how I can get it in my project? Thanks.

**Comment by patrick-rodgers** [comment]
PR will be in the next release on April 5 as noted in the attached milestone. I just published a beta v1.3.2-0 you can use to test these changes and see if they fix your issue.

**Comment by patrick-rodgers** [comment]
Going to close this as fixed by #581, change will be in upcoming release on Friday April 5 1.3.2 and is currently available in the published beta [mentioned here](https://github.com/pnp/pnpjs/issues/574#issuecomment-476949671)
